### PR TITLE
Refacto: model_config

### PR DESCRIPTION
**Pull request by vviers**
Still TODO 👷🏼‍♂️ :
- ~fill the config file with the correct values from the March model~
- ~create the CLI / script that consumes this config file and runs the model~

### CODE REVIEW

**Comment by vviers** [comment]
To run the script :

Install the package (as always)
`pip install -e .`

and run 🙂 
`python -m predictsignauxfaibles`

File: predictsignauxfaibles/__main__.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    step = "[TEST]"
+    logging.info(f"{step} - Fetching test set")
+    conf.TEST_DATASET.fetch_data()
+
+    logging.info(f"{step} - Data preprocessing")
+    conf.TEST_DATASET.replace_missing_data().remove_na(
+        ignore=["time_til_outcome"]
+    ).remove_strong_signals()
+    conf.TEST_DATASET.data = run_pipeline(conf.TEST_DATASET.data, conf.TRANSFO_PIPELINE)
```
Comment (slebastard) [inline predictsignauxfaibles/__main__.py]: If the general idea is to have a test set built from a given timeframe, we may want to enforce that there are no common SIREN in the train and test sets, by doing something similar to what is being done in model_selection.make_sf_test_validate_splits:
```python
    train_siren_set = conf.TRAIN_DATASET.data["siren"].unique()
    test_data = conf.TEST_DATASET.data[~conf.TEST_DATASET.data["siren"].isin(train_siren_set)]
```

We could also think of multiple test sets, for instance to test both forward and backward transfers of prediction performance: if we train a model on data from 2016, we may want to know whether it performed well in 2017 and 2018, but also in 2014 and 2015. This is a little far-fetched and should perhaps not be part of the PR though :)

File: predictsignauxfaibles/__main__.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    logging.info(f"{step} - Testing on {len(conf.TEST_DATASET)} observations.")
+
+    balanced_accuracy = balanced_accuracy_score(
```
Comment (slebastard) [inline predictsignauxfaibles/__main__.py]: We may want to also store the F-beta score. This could be part of a future issue.

File: tests/unit/data_test.py [hunk 3]
```diff
@@ -28,42 +27,40 @@ def test_full_drop_na(fake_testing_dataset):
...
-    must drop 11 (NAs) and then 5 more (strong signals)
-    must reset index
+    Methods must return self in order to be chained
     """
-    fake_testing_dataset.prepare_data(cols_ignore_na=[])
-    assert len(fake_testing_dataset) == 9
-    fake_testing_dataset.prepare_data(remove_strong_signals=True)
-    assert len(fake_testing_dataset) == 4
-    assert (fake_testing_dataset.data.index == list(range(4))).all()
+    fake_testing_dataset.replace_missing_data(defaults_map={"my_feature": 0}).remove_na(
```
Comment (slebastard) [inline tests/unit/data_test.py:63]: Here are you only testing the fact that the output of the first preprocessing function can successfully be fedto the second function of the chain? Or are you also testing that functions `replace_missing_data` and `remove_na` are functioning together?
If you're not reassigning the output of this chain to `fake_testing_dataset`, the chain isn't doing anything, right?

File: predictsignauxfaibles/__main__.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        conf.PREDICT_DATASET.data, conf.TRANSFO_PIPELINE
+    )
+    logging.info(f"{step} - Predicting on {len(conf.PREDICT_DATASET)} observations.")
+    predictions = fit.predict_proba(conf.PREDICT_DATASET.data)
+    conf.PREDICT_DATASET.data["predicted_probability"] = predictions[:, 1]
+
+    logging.info(f"{step} - Exporting data to csv")
+    model_id = datetime.now().strftime("%Y%m%d-%H%M%S")
+    export_destination = f"predictions-{model_id}.csv"
+    conf.PREDICT_DATASET.data[["siren", "siret", "predicted_probability"]].to_csv(
```
Comment (slebastard) [inline predictsignauxfaibles/__main__.py]: For traceability purposes, another export that we can setup is a small json summing up the main metrics that were generated:
- training batch size
- test batch size
- prediction batch size
- accuracy and beta score metrics on test set
- accuracy and beta score metrics on train set

**Review by slebastard — COMMENTED** [review]
This is great work, and this modular and flexible pipeline system is really appealing!
We may want to add a couple of minor changes:
- exclude SIRENs from train in the test set
- add a bunch of parameters such as batch sizes (train/test/predict), model config filepath, perhaps train, and test timeframes?
- as a bonus, why not design something that allows for multiple test sets, for instance to test backward and forward transfer of prediction performance
- output a bunch of prediction metrics and basic parameters and stats, so that we can easily trace the performance of our models through time

File: predictsignauxfaibles/__main__.py (hunk 1 above)
Comment (vviers) [reply predictsignauxfaibles/__main__.py]: Yes, let's maybe implement it as a method of SFDataset ?

```python
SFDataset.remove_siren_in_common(other: SFDataset)
```

Or maybe let's keep it simple for now ?

```python
SFDataset.remove_siren(siren: list)
```

File: predictsignauxfaibles/__main__.py (hunk 2 above)
Comment (vviers) [reply predictsignauxfaibles/__main__.py]: Yes, let's merge this simplistic pipeline for now and work on it in another PR

File: predictsignauxfaibles/__main__.py (hunk 4 above)
Comment (vviers) [reply predictsignauxfaibles/__main__.py]: Yes ! We could also pickle the sklearn fitted model object if we want

File: tests/unit/data_test.py (hunk 3 above)
Comment (vviers) [reply tests/unit/data_test.py:63]: It is ! both `replace_missing_data` and `remove_na` work on the data in place

Returning `self` is just a little trick in order to be able to chain methods but it does not change how the function behaves

Comment (vviers) [reply tests/unit/data_test.py:63]: You _can_ catch the object returned by the method, but you don't _have_ to 

We act in_place _and_ return, which I realize now might be confusing and not a good practice...

File: predictsignauxfaibles/__main__.py (hunk 1 above)
Comment (slebastard) [reply predictsignauxfaibles/__main__.py]: `SFDataset.remove_siren(siren: list)` seems nice and canonic, I'll go for that.

File: tests/unit/data_test.py (hunk 3 above)
Comment (slebastard) [reply tests/unit/data_test.py:63]: Ok, that makes sense!
Although it's not best practice, the only risk that I see is someone thinking a dataset is not modified (for instance during an assertion) although it actually is.
We can work on that in a next issue.

File: predictsignauxfaibles/__main__.py [hunk 5]
```diff
@@ -40,6 +40,9 @@ def run():
     logging.info(f"{step} - Fetching test set")
     conf.TEST_DATASET.fetch_data()
 
+    train_siren_set = conf.TRAIN_DATASET.data["siren"].unique().tolist()
```
Comment (vviers) [inline predictsignauxfaibles/__main__.py]: `logging.info(f"{step} - Removing train siren from test set")`

File: predictsignauxfaibles/data.py [hunk 6]
```diff
@@ -204,6 +204,20 @@ def remove_na(self, ignore: list):
...
         return self
 
+    def remove_siren(self, siren_list: list):
+        """
+        Removes all observations corresponding to
+        one of the siren provided as input.
+        Args:
+            siren_list: a list of siren to be removed from the data.
+        """
+        orig_length = len(self.data.index)
```
Comment (vviers) [inline predictsignauxfaibles/data.py]: you can use `len(self)`

File: predictsignauxfaibles/data.py [hunk 7]
```diff
@@ -204,6 +204,20 @@ def remove_na(self, ignore: list):
...
+        self.data = self.data[~self.data["siren"].isin(siren_list)]
+        post_length = len(self.data.index)
```
Comment (vviers) [inline predictsignauxfaibles/data.py]: itou : `len(self)`

File: predictsignauxfaibles/data.py (hunk 6 above)
Comment (slebastard) [reply predictsignauxfaibles/data.py]: My bad, didn't see method len() specialised here

**Comment by slebastard** [comment]
Added some functionalities:
- a step was added to remove SIREN from the train dataset in the test dataset
- predictsignauxfaibles/__main__.py now uses argparse to take arguments for variables that a non data-scientist user may want to play with when training models and outputing lists.
Models can now be trained like this: `python3 predictsignauxfaibles --model_name my_model --config_file my_model_config --train_from 01/01/2015 --train_to 01/09/2015`. When specifying a custom model and/or configuration file, we use the specific parameters declared in configuration file my_model/my_model_config. Those parameters can be overwritten à la carte using specific options:
    - "model_name" and "config_file" allow to use a different model and/or model configuration
    - oversampling
    - train_spl_size
    - train_from, train_to
    - test_spl_size
    - test_from, test_to
    - predict_on
- __main__.run() stores model stats in a dictionary and saves them in csv format
@vviers can you let me know what you think?

File: predictsignauxfaibles/__main__.py [hunk 8]
```diff
@@ -1,80 +1,207 @@
...
+def load_conf(args):  # pylint: disable=redefined-outer-name
+    """
+    Loads a model configuration from a argparse.Namespace
+    containing a model name and a configuration filename
+    Args:
+        conf_args: a argparse.Namespace object containing attributes
+            model_name
+            model_conf_path
+    """
+    conf_filepath = path.join("models", args.model_name, args.model_conf_path + ".py")
```
Comment (vviers) [inline predictsignauxfaibles/__main__.py]: IMO a `model_name` should only ever have 1 conf file (always named `model_conf.py`)

File: models/default/model_conf.py [hunk 9]
```diff
@@ -110,30 +109,13 @@
...
 TRAIN_OVERSAMPLING = 0.2
 
-TRAIN_DATASET = OversampledSFDataset(
-    TRAIN_OVERSAMPLING,
-    date_min=TRAIN_FROM,
-    date_max=TRAIN_TO,
-    fields=VARIABLES,
-    sample_size=TRAIN_SAMPLE_SIZE,
-)
-
```
Comment (vviers) [inline models/default/model_conf.py]: Why do you not want to keep dataset configuration in the model_conf file ?

File: predictsignauxfaibles/__main__.py [hunk 10]
```diff
@@ -1,80 +1,207 @@
...
+
+
+def run(args):  # pylint: disable=redefined-outer-name
     """
     Run model
     """
     logging.info(f"Running Model {conf.MODEL_ID} (commit {conf.MODEL_GIT_SHA})")
 
     step = "[TRAIN]"
+    TRAIN_DATASET = OversampledSFDataset(
```
Comment (vviers) [inline predictsignauxfaibles/__main__.py]: if we create the datasets in the same module they are used, they should be lowercased (`train_dataset`) ; ~i'm actually surprised that pylint didn't catch this~ it was ignored 🙃

File: predictsignauxfaibles/__main__.py [hunk 11]
```diff
@@ -1,80 +1,207 @@
...
     logging.info(f"{step} - Exporting data to csv")
     model_id = datetime.now().strftime("%Y%m%d-%H%M%S")
     export_destination = f"predictions-{model_id}.csv"
-    conf.PREDICT_DATASET.data[["siren", "siret", "predicted_probability"]].to_csv(
+    PREDICT_DATASET.data[["siren", "siret", "predicted_probability"]].to_csv(
         export_destination, index=False
     )
 
 
+parser = argparse.ArgumentParser("main.py", description="Run model prediction")
```
Comment (vviers) [inline predictsignauxfaibles/__main__.py:158]: I'm a little new to argparse (i usually use `Click` to make CLI) so sorry if that's a dumb question but should it not be `__main__.py` instead of `main.py` ?

File: predictsignauxfaibles/__main__.py [hunk 12]
```diff
@@ -40,13 +41,30 @@ def load_conf(args):  # pylint: disable=redefined-outer-name
...
+def evaluate(model, df):  # To be turned into a SFModel method when refactoring models
+    """
+    Returns evaluation metrics of model evaluated on df
+    Args:
+        model: a sklearn-like model with a predict method
+        df: dataset
+    """
+    balanced_accuracy = balanced_accuracy_score(df["outcome"], model.predict(df))
+    fbeta = fbeta_score(df["outcome"], model.predict(df), beta=conf.EVAL_BETA)
+    return balanced_accuracy, fbeta
```
Comment (vviers) [inline predictsignauxfaibles/__main__.py]: `return {"balanced_accuracy": balanced_accuracy, "fbeta": fbeta}`

since nothing in the function signature tells us which metrics are returned, better make it explicit

File: predictsignauxfaibles/__main__.py [hunk 13]
```diff
@@ -40,13 +41,30 @@ def load_conf(args):  # pylint: disable=redefined-outer-name
...
+
+def run(
+    args,
+):  # pylint: disable=redefined-outer-name,too-many-statements,too-many-locals
     """
     Run model
     """
     logging.info(f"Running Model {conf.MODEL_ID} (commit {conf.MODEL_GIT_SHA})")
+    stats = vars(args)
+    stats["run_on"] = datetime.datetime()
```
Comment (vviers) [inline predictsignauxfaibles/__main__.py]: let's maybe call this `model_stats` so that it's slightly more explicit

File: predictsignauxfaibles/__main__.py (hunk 8 above)
Comment (vviers) [reply predictsignauxfaibles/__main__.py]: ✅

File: models/default/model_conf.py (hunk 9 above)
Comment (vviers) [reply models/default/model_conf.py]: As discussed together, that allows for passing params in the CLI — ✅

File: predictsignauxfaibles/__main__.py (hunk 10 above)
Comment (vviers) [reply predictsignauxfaibles/__main__.py]: ✅

File: predictsignauxfaibles/__main__.py (hunk 12 above)
Comment (vviers) [reply predictsignauxfaibles/__main__.py]: ✅

File: predictsignauxfaibles/__main__.py (hunk 13 above)
Comment (vviers) [reply predictsignauxfaibles/__main__.py]: ✅

**Review by slebastard — APPROVED** [review]
LGTM! Merging
